File: leetcode/go/000101.symmetric_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return isMirror(root.Left, root.Right)
}

func isMirror(a, b *TreeNode) bool {
	if a == nil || b == nil {
		return a == b
	}

	return a.Val == b.Val && isMirror(a.Left, b.Right) && isMirror(a.Right, b.Left)
}

// Time complexity: O(n)
// Space complexity: O(h)

File: leetcode/typescript/000101.symmetric-tree.ts
// Definition for a binary tree node.
class TreeNode {
	val: number;
	left: TreeNode | null;
	right: TreeNode | null;
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = val === undefined ? 0 : val;
		this.left = left === undefined ? null : left;
		this.right = right === undefined ? null : right;
	}
}

// Time complexity: O(n)
// Space complexity: O(h)
function isSymmetric(root: TreeNode | null): boolean {
	if (!root) return true;
	return isMirror(root.left, root.right);
}

function isMirror(left: TreeNode | null, right: TreeNode | null): boolean {
	if (!left || !right) return left === right;
	if (!(left.val === right.val)) return false;
	return isMirror(left.left, right.right) && isMirror(left.right, right.left);
}
